x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/update/update_gaps.ts

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Logger, ISavedObjectsRepository } from '@kbn/core/server';
+import { IEventLogClient, IEventLogger } from '@kbn/event-log-plugin/server';
+import { ActionsClient } from '@kbn/actions-plugin/server';
+import { BackfillClient } from '../../../backfill_client/backfill_client';
+import { AlertingEventLogger } from '../../alerting_event_logger/alerting_event_logger';
+import { findGaps } from '../find_gaps';
+import { findGapById } from '../find_gap_by_id';
+import { Gap } from '../gap';
+import { gapStatus } from '../../../../common/constants';
+import { BackfillSchedule } from '../../../application/backfill/result/types';
+import { adHocRunStatus } from '../../../../common/constants';
+import { calculateGapStateFromAllBackfills } from './calculate_gaps_state';
+import { updateGapFromSchedule } from './update_gap_from_schedule';
+
+interface UpdateGapsParams {
+  ruleId: string;
+  backfillSchedule?: BackfillSchedule[];
+  start: Date;
+  end: Date;
+  eventLogger?: IEventLogger;
+  eventLogClient: IEventLogClient;
+  logger: Logger;
+  savedObjectsRepository: ISavedObjectsRepository;
+  shouldRefetchAllBackfills?: boolean;
+  backfillClient: BackfillClient;
+  actionsClient: ActionsClient;
+}
+
+const CONFLICT_STATUS_CODE = 409;
+const MAX_RETRIES = 3;
+
+const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+const updateSingleGap = async (
+  gap: Gap,
+  {
+    backfillSchedule,
+    savedObjectsRepository,
+    shouldRefetchAllBackfills,
+    backfillClient,
+    actionsClient,
+    alertingEventLogger,
+    logger,
+    ruleId,
+    eventLogClient,
+  }: {
+    backfillSchedule?: BackfillSchedule[];
+    savedObjectsRepository: ISavedObjectsRepository;
+    shouldRefetchAllBackfills?: boolean;
+    backfillClient: BackfillClient;
+    actionsClient: ActionsClient;
+    alertingEventLogger: AlertingEventLogger;
+    logger: Logger;
+    ruleId: string;
+    eventLogClient: IEventLogClient;
+  }
+) => {
+  let currentGap = gap;
+  let retryCount = 0;
+
+  while (retryCount <= MAX_RETRIES) {
+    try {
+      // If we're retrying, refetch the latest gap state
+      if (retryCount > 0 && currentGap.internalFields) {
+        const updatedGap = await findGapById({
```

Comment:
if we know the gap ID, can we just do a `GET` on the document instead of a `find`?

Reply:
Change to in new version to get list of failed gaps by id